File: golang/UnionFind/unionfind.go
package main

import (
	"fmt"
)

// UnionFind 并查集接口（最优版本：路径压缩 + 按秩合并）
type UnionFind struct {
	parent []int
	rank   []int
}

// NewUnionFind 创建一个新的并查集
func NewUnionFind(size int) *UnionFind {
	parent := make([]int, size)
	rank := make([]int, size)
	for i := 0; i < size; i++ {
		parent[i] = i
		rank[i] = 1
	}
	return &UnionFind{
		parent: parent,
		rank:   rank,
	}
}

func (uf *UnionFind) Size() int {
	return len(uf.parent)
}

// IsConnected 检查两个元素是否连接
func (uf *UnionFind) IsConnected(p, q int) bool {
	return uf.find(p) == uf.find(q)
}

// UnionElements 合并两个元素
func (uf *UnionFind) UnionElements(p, q int) {
	pRoot := uf.find(p)
	qRoot := uf.find(q)

	if pRoot == qRoot {
		return
	}

	// 按秩合并 - 尽量保持树的深度一致
	if uf.rank[pRoot] > uf.rank[qRoot] {
		uf.parent[qRoot] = pRoot
	} else if uf.rank[pRoot] < uf.rank[qRoot] {
		uf.parent[pRoot] = qRoot
	} else {
		uf.parent[pRoot] = qRoot
		uf.rank[qRoot]++
	}
}

// find 查找根节点，带路径压缩
func (uf *UnionFind) find(p int) int {
	if p < 0 || p >= len(uf.parent) {
		panic("Index out of range")
	}

	// 路径压缩
	if p != uf.parent[p] {
		uf.parent[p] = uf.find(uf.parent[p])
	}
	return uf.parent[p]
}

// CountSets 统计集合数量
func (uf *UnionFind) CountSets() int {
	roots := make(map[int]bool)
	for i := 0; i < len(uf.parent); i++ {
		roots[uf.find(i)] = true
	}
	return len(roots)
}

func (uf *UnionFind) String() string {
	return fmt.Sprintf("UnionFind(elements=%d, sets=%d)", len(uf.parent), uf.CountSets())
}

// UnionFindV1 第一版并查集：基于数组，时间复杂度O(n)
type UnionFindV1 struct {
	data [][]int
}

// NewUnionFindV1 创建第一版并查集
func NewUnionFindV1(size int) *UnionFindV1 {
	data := make([][]int, size)
	for i := 0; i < size; i++ {
		data[i] = []int{i}
	}
	return &UnionFindV1{data: data}
}

func (uf *UnionFindV1) IsConnected(p, q int) bool {
	return uf.findId(p) == uf.findId(q)
}

func (uf *UnionFindV1) findId(p int) int {
	for i, s := range uf.data {
		for _, v := range s {
			if v == p {
				return i
			}
		}
	}
	return -1
}

func (uf *UnionFindV1) UnionElements(p, q int) {
	pId := uf.findId(p)
	qId := uf.findId(q)

	if pId == qId {
		return
	}

	// 将q的集合合并到p的集合
	uf.data[pId] = append(uf.data[pId], uf.data[qId]...)
	// 删除q的集合
	uf.data = append(uf.data[:qId], uf.data[qId+1:]...)
}

// UnionFindV2 第二版并查集：Quick Find
type UnionFindV2 struct {
	id []int
}

// NewUnionFindV2 创建第二版并查集
func NewUnionFindV2(size int) *UnionFindV2 {
	id := make([]int, size)
	for i := 0; i < size; i++ {
		id[i] = i
	}
	return &UnionFindV2{id: id}
}

func (uf *UnionFindV2) IsConnected(p, q int) bool {
	return uf.id[p] == uf.id[q]
}

func (uf *UnionFindV2) find(p int) int {
	if p < 0 || p >= len(uf.id) {
		panic("Index out of range")
	}
	return uf.id[p]
}

func (uf *UnionFindV2) UnionElements(p, q int) {
	pId := uf.find(p)
	qId := uf.find(q)

	if pId == qId {
		return
	}

	// 将所有id[p]的元素改为id[q]
	for i := 0; i < len(uf.id); i++ {
		if uf.id[i] == pId {
			uf.id[i] = qId
		}
	}
}

// UnionFindV3 第三版并查集：Quick Union
type UnionFindV3 struct {
	parent []int
}

// NewUnionFindV3 创建第三版并查集
func NewUnionFindV3(size int) *UnionFindV3 {
	parent := make([]int, size)
	for i := 0; i < size; i++ {
		parent[i] = i
	}
	return &UnionFindV3{parent: parent}
}

func (uf *UnionFindV3) IsConnected(p, q int) bool {
	return uf.find(p) == uf.find(q)
}

func (uf *UnionFindV3) find(p int) int {
	if p < 0 || p >= len(uf.parent) {
		panic("Index out of range")
	}
	for p != uf.parent[p] {
		p = uf.parent[p]
	}
	return p
}

func (uf *UnionFindV3) UnionElements(p, q int) {
	pRoot := uf.find(p)
	qRoot := uf.find(q)

	if pRoot == qRoot {
		return
	}

	uf.parent[pRoot] = qRoot
}

// UnionFindV4 第四版并查集：基于size的优化
type UnionFindV4 struct {
	parent []int
	size   []int
}

// NewUnionFindV4 创建第四版并查集
func NewUnionFindV4(size int) *UnionFindV4 {
	parent := make([]int, size)
	sz := make([]int, size)
	for i := 0; i < size; i++ {
		parent[i] = i
		sz[i] = 1
	}
	return &UnionFindV4{
		parent: parent,
		size:   sz,
	}
}

func (uf *UnionFindV4) IsConnected(p, q int) bool {
	return uf.find(p) == uf.find(q)
}

func (uf *UnionFindV4) find(p int) int {
	if p < 0 || p >= len(uf.parent) {
		panic("Index out of range")
	}
	for p != uf.parent[p] {
		p = uf.parent[p]
	}
	return p
}

func (uf *UnionFindV4) UnionElements(p, q int) {
	pRoot := uf.find(p)
	qRoot := uf.find(q)

	if pRoot == qRoot {
		return
	}

	// 基于size的优化，将小树连接到大树
	if uf.size[pRoot] < uf.size[qRoot] {
		uf.parent[pRoot] = qRoot
		uf.size[qRoot] += uf.size[pRoot]
	} else {
		uf.parent[qRoot] = pRoot
		uf.size[pRoot] += uf.size[qRoot]
	}
}

// UnionFindV5 第五版并查集：路径压缩优化
type UnionFindV5 struct {
	parent []int
	rank   []int
}

// NewUnionFindV5 创建第五版并查集
func NewUnionFindV5(size int) *UnionFindV5 {
	parent := make([]int, size)
	rank := make([]int, size)
	for i := 0; i < size; i++ {
		parent[i] = i
		rank[i] = 1
	}
	return &UnionFindV5{
		parent: parent,
		rank:   rank,
	}
}

func (uf *UnionFindV5) IsConnected(p, q int) bool {
	return uf.find(p) == uf.find(q)
}

func (uf *UnionFindV5) find(p int) int {
	if p < 0 || p >= len(uf.parent) {
		panic("Index out of range")
	}

	// 路径压缩
	if p != uf.parent[p] {
		uf.parent[p] = uf.find(uf.parent[p])
	}
	return uf.parent[p]
}

func (uf *UnionFindV5) UnionElements(p, q int) {
	pRoot := uf.find(p)
	qRoot := uf.find(q)

	if pRoot == qRoot {
		return
	}

	if uf.rank[pRoot] < uf.rank[qRoot] {
		uf.parent[pRoot] = qRoot
	} else if uf.rank[pRoot] > uf.rank[qRoot] {
		uf.parent[qRoot] = pRoot
	} else {
		uf.parent[pRoot] = qRoot
		uf.rank[qRoot]++
	}
}

func main() {
	fmt.Println("=== Go语言并查集示例 ===")

	// 测试最优版本
	fmt.Println("\n--- UnionFind V6 (路径压缩 + 按秩合并) ---")
	uf := NewUnionFind(10)
	fmt.Println(uf)

	uf.UnionElements(0, 1)
	uf.UnionElements(2, 3)
	uf.UnionElements(1, 2)
	uf.UnionElements(4, 5)
	uf.UnionElements(6, 7)

	fmt.Printf("合并后:\n")
	fmt.Printf("IsConnected(0, 3): %t\n", uf.IsConnected(0, 3))
	fmt.Printf("IsConnected(0, 4): %t\n", uf.IsConnected(0, 4))
	fmt.Printf("CountSets: %d\n", uf.CountSets())

	// 测试V2版本
	fmt.Println("\n--- UnionFind V2 (Quick Find) ---")
	uf2 := NewUnionFindV2(10)
	uf2.UnionElements(0, 1)
	uf2.UnionElements(2, 3)
	uf2.UnionElements(1, 2)
	fmt.Printf("IsConnected(0, 3): %t\n", uf2.IsConnected(0, 3))

	// 测试V3版本
	fmt.Println("\n--- UnionFind V3 (Quick Union) ---")
	uf3 := NewUnionFindV3(10)
	uf3.UnionElements(0, 1)
	uf3.UnionElements(2, 3)
	uf3.UnionElements(1, 2)
	fmt.Printf("IsConnected(0, 3): %t\n", uf3.IsConnected(0, 3))

	fmt.Println("\n=== 示例程序结束 ===")
}

File: python/unionfind.py
"""
Union Find (Disjoint Set) Implementation in Python
Based on Java UnionFind implementation
"""

from typing import List


class UnionFind:
    """Union Find with path compression and union by rank (optimized version)"""

    def __init__(self, size: int):
        self._parent: List[int] = list(range(size))
        self._rank: List[int] = [1] * size

    def size(self) -> int:
        """Get number of elements"""
        return len(self._parent)

    def is_connected(self, p: int, q: int) -> bool:
        """Check if two elements are connected"""
        return self.find(p) == self.find(q)

    def union_elements(self, p: int, q: int) -> None:
        """Union two elements"""
        p_root = self.find(p)
        q_root = self.find(q)

        if p_root == q_root:
            return

        # Union by rank - attach smaller tree to larger tree
        if self._rank[p_root] > self._rank[q_root]:
            self._parent[q_root] = p_root
        elif self._rank[p_root] < self._rank[q_root]:
            self._parent[p_root] = q_root
        else:
            self._parent[p_root] = q_root
            self._rank[q_root] += 1

    def find(self, p: int) -> int:
        """Find root of element with path compression"""
        if p < 0 or p >= len(self._parent):
            raise IndexError("Index out of range")

        # Path compression
        if p != self._parent[p]:
            self._parent[p] = self.find(self._parent[p])
        return self._parent[p]

    def count_sets(self) -> int:
        """Count number of disjoint sets"""
        roots = set()
        for i in range(len(self._parent)):
            roots.add(self.find(i))
        return len(roots)

    def get_set_sizes(self) -> dict[int, int]:
        """Get size of each set"""
        set_sizes: dict[int, int] = {}
        for i in range(len(self._parent)):
            root = self.find(i)
            set_sizes[root] = set_sizes.get(root, 0) + 1
        return set_sizes

    def __str__(self) -> str:
        return f"UnionFind(elements={len(self._parent)}, sets={self.count_sets()})"

    def __repr__(self) -> str:
        return self.__str__()


class UnionFindV1:
    """Union Find Version 1: Array-based with path copying (O(n))"""

    def __init__(self, size: int):
        self._data: List[List[int]] = [[i] for i in range(size)]

    def size(self) -> int:
        return len(self._data)

    def is_connected(self, p: int, q: int) -> bool:
        if p < 0 or p >= len(self._data) or q < 0 or q >= len(self._data):
            raise IndexError("Index out of range")
        return self._find_id(p) == self._find_id(q)

    def _find_id(self, p: int) -> int:
        """Find which set element p belongs to"""
        for i, s in enumerate(self._data):
            if p in s:
                return i
        return -1

    def union_elements(self, p: int, q: int) -> None:
        """Union two elements"""
        p_id = self._find_id(p)
        q_id = self._find_id(q)

        if p_id == q_id:
            return

        # Merge q's set into p's set
        self._data[p_id].extend(self._data[q_id])
        self._data.pop(q_id)

    def find(self, p: int) -> int:
        """Find root (set identifier)"""
        return self._find_id(p)


class UnionFindV2:
    """Union Find Version 2: Quick-find with array id (O(n))"""

    def __init__(self, size: int):
        self._id: List[int] = list(range(size))

    def size(self) -> int:
        return len(self._id)

    def is_connected(self, p: int, q: int) -> bool:
        if p < 0 or p >= len(self._id) or q < 0 or q >= len(self._id):
            raise IndexError("Index out of range")
        return self._id[p] == self._id[q]

    def find(self, p: int) -> int:
        """Find the set identifier"""
        if p < 0 or p >= len(self._id):
            raise IndexError("Index out of range")
        return self._id[p]

    def union_elements(self, p: int, q: int) -> None:
        """Union two elements"""
        p_id = self.find(p)
        q_id = self.find(q)

        if p_id == q_id:
            return

        # Change all elements with id[p] to id[q]
        for i in range(len(self._id)):
            if self._id[i] == p_id:
                self._id[i] = q_id


class UnionFindV3:
    """Union Find Version 3: Quick-union with tree structure (O(n))"""

    def __init__(self, size: int):
        self._parent: List[int] = list(range(size))

    def size(self) -> int:
        return len(self._parent)

    def is_connected(self, p: int, q: int) -> bool:
        return self.find(p) == self.find(q)

    def find(self, p: int) -> int:
        """Find root of element"""
        if p < 0 or p >= len(self._parent):
            raise IndexError("Index out of range")
        while p != self._parent[p]:
            p = self._parent[p]
        return p

    def union_elements(self, p: int, q: int) -> None:
        """Union two elements"""
        p_root = self.find(p)
        q_root = self.find(q)

        if p_root == q_root:
            return

        self._parent[p_root] = q_root


class UnionFindV4:
    """Union Find Version 4: Union by rank optimization"""

    def __init__(self, size: int):
        self._parent: List[int] = list(range(size))
        self._rank: List[int] = [1] * size

    def size(self) -> int:
        return len(self._parent)

    def is_connected(self, p: int, q: int) -> bool:
        return self.find(p) == self.find(q)

    def find(self, p: int) -> int:
        """Find root of element"""
        if p < 0 or p >= len(self._parent):
            raise IndexError("Index out of range")
        while p != self._parent[p]:
            p = self._parent[p]
        return p

    def union_elements(self, p: int, q: int) -> None:
        """Union two elements with rank optimization"""
        p_root = self.find(p)
        q_root = self.find(q)

        if p_root == q_root:
            return

        # Union by rank
        if self._rank[p_root] < self._rank[q_root]:
            self._parent[p_root] = q_root
        elif self._rank[p_root] > self._rank[q_root]:
            self._parent[q_root] = p_root
        else:
            self._parent[p_root] = q_root
            self._rank[q_root] += 1


class UnionFindV5:
    """Union Find Version 5: Path compression optimization"""

    def __init__(self, size: int):
        self._parent: List[int] = list(range(size))

    def size(self) -> int:
        return len(self._parent)

    def is_connected(self, p: int, q: int) -> bool:
        return self.find(p) == self.find(q)

    def find(self, p: int) -> int:
        """Find root of element with path compression"""
        if p < 0 or p >= len(self._parent):
            raise IndexError("Index out of range")
        if p != self._parent[p]:
            self._parent[p] = self.find(self._parent[p])
        return self._parent[p]

    def union_elements(self, p: int, q: int) -> None:
        """Union two elements"""
        p_root = self.find(p)
        q_root = self.find(q)

        if p_root == q_root:
            return

        self._parent[p_root] = q_root


if __name__ == "__main__":
    # Test UnionFind (optimized version)
    print("Testing UnionFind (V6 - optimized):")
    uf = UnionFind(10)
    print(uf)

    uf.union_elements(0, 1)
    uf.union_elements(2, 3)
    uf.union_elements(1, 2)
    uf.union_elements(4, 5)
    uf.union_elements(6, 7)

    print(f"After unions:")
    print(f"Is 0 connected to 3: {uf.is_connected(0, 3)}")
    print(f"Is 0 connected to 4: {uf.is_connected(0, 4)}")
    print(f"Count sets: {uf.count_sets()}")
    print(f"Set sizes: {uf.get_set_sizes()}")

    # Test UnionFindV2
    print("\nTesting UnionFindV2 (Quick-find):")
    uf2 = UnionFindV2(10)
    uf2.union_elements(0, 1)
    uf2.union_elements(2, 3)
    uf2.union_elements(1, 2)
    print(f"Is 0 connected to 3: {uf2.is_connected(0, 3)}")

    # Test UnionFindV3
    print("\nTesting UnionFindV3 (Quick-union):")
    uf3 = UnionFindV3(10)
    uf3.union_elements(0, 1)
    uf3.union_elements(2, 3)
    uf3.union_elements(1, 2)
    print(f"Is 0 connected to 3: {uf3.is_connected(0, 3)}")
